extensions/src/main/java/org/pentaho/platform/plugin/services/email/EmailConfiguration.java

```diff
@@ -202,7 +288,12 @@ public boolean equals( final Object obj ) {
         && ObjectUtils.equals( this.getSmtpPort(), that.getSmtpPort() )
         && isEquals( this.defaultFrom, that.defaultFrom ) && isEquals( this.fromName, that.fromName )
         && isEquals( this.smtpHost, that.smtpHost ) && isEquals( this.smtpProtocol, that.smtpProtocol )
-        && isEquals( this.userId, that.userId ) && isEquals( this.password, that.password ) );
+        && isEquals( this.userId, that.userId ) && isEquals( this.password, that.password )
+        && isEquals( this.authMechanism, that.authMechanism ) && isEquals( this.clientId, that.clientId )
+        && isEquals( this.clientSecret, that.clientSecret ) && isEquals( this.tokenUrl, that.tokenUrl )
+        && isEquals( this.scope, that.scope ) && isEquals( this.grantType, that.grantType )
+        && isEquals( this.refreshToken, that.refreshToken ) && isEquals( this.authorizationCode, that.authorizationCode )
+        && isEquals( this.redirectUri, that.redirectUri ) );
```

Comment:
You need to override EmailConfiguration#hashcode() (to test the same variables in the equals)

In java Objects have a contract:
    _If ObjA.equals(ObjB) == true then ObjA.hashcode() == ObjB.hashcode()_

It this relationship isn't in place, then if you use a map/set, then exists() will always return false, and you could have endlesss copies of the "same" object because default hascode() goes off java memory location.

Original author(s) should have caught this.

https://docs.oracle.com/javase/7/docs/api/java/lang/Object.html#hashCode()
second bullet
- _"If two objects are equal according to the equals(Object) method, then calling the hashCode method on each of the two objects must produce the same integer result."_
 
 There are 3 ways to implement hashcode:
 I would advise go with option 2 or 3, but all 3 are technically correct.
 
 option 1: hand code everything (don't advise)
```
@Override
       public int hashCode() {
              final int prime = 31;
              int result = 1;
              result = prime * result
                           + ((exchange == null) ? 0 : exchange.hashCode());
              result = prime * result + (isRestricted ? 1231 : 1237);
              result = prime * result + (int) (lotSize ^ (lotSize >>> 32));
```
Read more: https://javarevisited.blogspot.com/2011/10/override-hashcode-in-java-example.html#ixzz7rEDtgj9b

Option 2: Use Apaches ObjectUtils.hashCode (good approach)
```
@Override
public int hashCode() {
  int hash = 17;
  hash = hash * 31 + ObjectUtils.hashCode(personId);
  hash = hash * 31 + ObjectUtils.hashCode(firstName);
```
read mode: https://www.tabnine.com/code/java/methods/org.apache.commons.lang.ObjectUtils/hashCode?snippet=5ce7084e7e034400044184cb

option 3: Apache's HashCodeBuilder
```
  @Override
    public int hashCode() {
        return new HashCodeBuilder().append(id).append(name).toHashCode();
    }
```
read more: https://kodejava.org/how-to-implement-the-hashcode-and-equals-method-using-apache-commons/